### ISSUE TITLE
global name conflicts with class member

### ISSUE THREAD

**Issue by Akuli** [microsoft/pyright#2443]
**Describe the bug**

If you have a method named `list` in a class, you cannot use the built-in `list` type.

**To Reproduce**

```python
from __future__ import annotations
class Foo:
    def list(self) -> str: ...
    def blah(self) -> list[int]: ...
```

**Expected behavior**
No error. (consistent with mypy)

**VS Code extension or command-line**
```
akuli@akuli-desktop:~/typeshed$ npx pyright@1.1.175 asd.py
npx: installed 1 in 5.869s
Loading configuration file at /home/akuli/typeshed/pyrightconfig.json
Python version 2.7 from interpreter is unsupported
Assuming Python platform Linux
stubPath /home/akuli/typeshed/typings is not a valid directory.
Searching for source files
Found 1 source file
/home/akuli/typeshed/asd.py
  /home/akuli/typeshed/asd.py:3:23 - error: Expected class type but received "(self: Foo) -> str" (reportGeneralTypeIssues)
  /home/akuli/typeshed/asd.py:3:23 - error: Object of type "(self: Foo) -> str" is not subscriptable (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos 
Completed in 0.753sec
```

**Additional context**
We occasionally run into this in typeshed. I'm not sure if it's intentional.

**Comment by erictraut** [comment]
This is correct behavior. If you are using `from __future__ import annotations`, all type annotations are treated by a type checker as forward-declared lookups, just as if you had enclosed them in quotes. A forward-declared lookup does not take into account code flow; it uses the symbol table of the current scope. In your case, that means it uses the symbol `list` within the class `Foo`.

There are two ways you can fix this in your code:

1. Remove `from __future__ import annotations`. Pyright will then apply code flow ordering to resolve the `list` symbol.
2. Avoid the namespace conflict by doing the following:

```python
from builtins import list as _list

class Foo:
    def list(self) -> str: ...
    def blah(self) -> _list[int]: ...
```

or 

```python
_list = list

class Foo:
    def list(self) -> str: ...
    def blah(self) -> _list[int]: ...
```

I would recommend option 2 because `from __future__ import annotations` is likely to become the default mode in Python 3.11.

**Comment by Akuli** [comment]
Thanks. I was somehow imagining that type annotations get evaluated as if they were inside the function body, where `list` means a built-in, even if the class has a member named `list`.
